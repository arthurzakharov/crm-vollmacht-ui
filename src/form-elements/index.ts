export * from "./checkbox";
export * from "./input";
export * from "./input-checkbox";
export * from "./input-radio";
export * from "./input-text";
export * from "./navigation-button";
export * from "./radio";
