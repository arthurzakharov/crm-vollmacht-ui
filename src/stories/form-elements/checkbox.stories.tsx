import { Meta, StoryObj } from "@storybook/react";
import { Checkbox } from "../../form-elements";

type Story = StoryObj<typeof Checkbox>;

const meta: Meta<typeof Checkbox> = {
  title: "FORM ELEMENTS/Checkbox",
  component: Checkbox,
  args: {
    value: true,
    focused: false,
    disabled: false,
  },
};

export const Checked: Story = {
  name: "Checked",
  args: {
    value: true,
  },
};

export const NotChecked: Story = {
  name: "Not Checked",
  args: {
    value: false,
  },
};

export const Focused: Story = {
  name: "Focused",
  args: {
    focused: true,
  },
};

export const NotFocused: Story = {
  name: "Not Focused",
  args: {
    focused: false,
  },
};

export const Disabled: Story = {
  name: "Disabled",
  args: {
    disabled: true,
  },
};

export const Enabled: Story = {
  name: "Enabled",
  args: {
    disabled: false,
  },
};

export const DefaultValues: Story = {
  name: "Default Values",
  args: {
    focused: undefined,
    disabled: undefined,
  },
};

export default meta;
