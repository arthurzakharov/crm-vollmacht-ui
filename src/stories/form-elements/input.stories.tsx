import React, { useState } from "react";
import { Meta, StoryObj } from "@storybook/react";
import { fn, within, expect, userEvent } from "@storybook/test";
import { Input } from "../../form-elements";
import MaxWidthDecorator from "../decorators/max-width-decorator";

type Story = StoryObj<typeof Input>;

const meta: Meta<typeof Input> = {
  title: "FORM ELEMENTS/Input",
  component: Input,
  parameters: {
    MaxWidthDecorator: 280,
  },
  decorators: [MaxWidthDecorator],
  render: (args) => {
    const [value, setValue] = useState<string>(args.value);
    const onChange = (v: string) => {
      args.onChange(v);
      setValue(v);
    };
    return <Input {...args} value={value} onChange={onChange} />;
  },
};

export const Demo: Story = {
  name: "Demo",
  args: {
    value: "",
    name: "input",
    color: "primary",
    masked: true,
    status: "neutral",
    disabled: false,
    placeholder: "TT/MM/JJJJ",
    maxLength: 12,
    type: "text",
    onChange: fn(),
    onFocus: fn(),
    onBlur: fn(),
  },
};

export const ColorPrimary: Story = {
  name: "Color Primary",
  args: {
    ...Demo.args,
    value: "19/03/1988",
    color: "primary",
  },
};

export const ColorSecondary: Story = {
  name: "Color Secondary",
  args: {
    ...Demo.args,
    value: "19/03/1988",
    color: "secondary",
  },
};

export const StatusNeutral: Story = {
  name: "Status Neutral",
  args: {
    ...Demo.args,
    value: "19/03/1988",
    status: "neutral",
  },
};

export const StatusSuccess: Story = {
  name: "Status Success",
  args: {
    ...Demo.args,
    value: "19/03/1988",
    status: "success",
  },
};

export const StatusError: Story = {
  name: "Status Error",
  args: {
    ...Demo.args,
    value: "19/03/1988",
    status: "error",
  },
};

export const Masked: Story = {
  name: "Masked",
  args: {
    ...Demo.args,
    masked: true,
  },
  play: async ({ args, canvasElement, step }) => {
    const input = within(canvasElement).getByRole("textbox");
    await step("Focus in and out", async () => {
      await userEvent.tab();
      await userEvent.tab();
      await expect(args.onFocus).toHaveBeenNthCalledWith(1);
      await expect(args.onBlur).toHaveBeenNthCalledWith(1);
      await expect(args.onChange).toHaveBeenNthCalledWith(1, "__/__/____");
      await expect(args.onChange).toHaveBeenNthCalledWith(2, "");
    });
    await step("Type", async () => {
      await userEvent.type(input, "12122000");
      await expect(args.onChange).toHaveBeenNthCalledWith(3, "__/__/____");
      await expect(args.onChange).toHaveBeenNthCalledWith(4, "1_/__/____");
      await expect(args.onChange).toHaveBeenNthCalledWith(5, "12/__/____");
      await expect(args.onChange).toHaveBeenNthCalledWith(6, "12/1_/____");
      await expect(args.onChange).toHaveBeenNthCalledWith(7, "12/12/____");
      await expect(args.onChange).toHaveBeenNthCalledWith(8, "12/12/2___");
      await expect(args.onChange).toHaveBeenNthCalledWith(9, "12/12/20__");
      await expect(args.onChange).toHaveBeenNthCalledWith(10, "12/12/200_");
      await expect(args.onChange).toHaveBeenNthCalledWith(11, "12/12/2000");
      await userEvent.tab();
    });
    await step("Check that caret always returns to first not entered number", async () => {
      await userEvent.tab();
      await userEvent.type(input, "{backspace}");
      await userEvent.type(input, "{backspace}");
      await userEvent.type(input, "{backspace}");
      await expect(args.onChange).toHaveBeenNthCalledWith(14, "12/12/2___");
      await userEvent.tab();
      await userEvent.click(input);
      await userEvent.keyboard("8");
      await expect(args.onChange).toHaveBeenNthCalledWith(15, "12/12/28__");
      await userEvent.tab();
    });
    await step("Press ArrowUp and ArrowDown does not move caret", async () => {
      await userEvent.click(input);
      await userEvent.keyboard("[ArrowUp]");
      await userEvent.keyboard("8");
      await expect(args.onChange).toHaveBeenNthCalledWith(16, "12/12/288_");
      await userEvent.type(input, "{backspace}");
      await expect(args.onChange).toHaveBeenNthCalledWith(17, "12/12/28__");
      await userEvent.keyboard("[ArrowDown]");
      await userEvent.keyboard("8");
      await expect(args.onChange).toHaveBeenNthCalledWith(18, "12/12/288_");
      await userEvent.type(input, "{backspace}");
      await userEvent.type(input, "{backspace}");
      await userEvent.type(input, "{backspace}");
      await expect(args.onChange).toHaveBeenNthCalledWith(21, "12/12/____");
      await userEvent.tab();
    });
    await step("Press ArrowLeft and ArrowRight move caret", async () => {
      await userEvent.click(input);
      await userEvent.keyboard("[ArrowRight]");
      await userEvent.keyboard("2");
      await expect(args.onChange).toHaveBeenNthCalledWith(22, "12/12/2___");
      await userEvent.keyboard("[ArrowLeft]");
      await userEvent.keyboard("[ArrowLeft]");
      await userEvent.keyboard("[ArrowLeft]");
      await userEvent.keyboard("[ArrowLeft]");
      await userEvent.keyboard("1");
      await expect(args.onChange).toHaveBeenNthCalledWith(23, "11/12/2___");
      await userEvent.keyboard("[ArrowRight]");
      await userEvent.keyboard("[ArrowRight]");
      await userEvent.keyboard("[ArrowRight]");
      await userEvent.keyboard("1");
      await expect(args.onChange).toHaveBeenNthCalledWith(24, "11/12/1___");
      await userEvent.keyboard("9");
      await userEvent.keyboard("9");
      await userEvent.keyboard("9");
      await expect(args.onChange).toHaveBeenNthCalledWith(27, "11/12/1999");
    });
  },
};

export const NotMasked: Story = {
  name: "NotMasked",
  args: {
    ...Demo.args,
    masked: false,
  },
  play: async ({ args, canvasElement, step }) => {
    const input = within(canvasElement).getByRole("textbox");
    await step("Focus in and out", async () => {
      await userEvent.tab();
      await userEvent.tab();
      await expect(args.onFocus).toHaveBeenNthCalledWith(1);
      await expect(args.onBlur).toHaveBeenNthCalledWith(1);
      await expect(args.onChange).not.toBeCalled();
    });
    await step("Type", async () => {
      await userEvent.type(input, "test");
      await expect(args.onChange).toHaveBeenNthCalledWith(1, "t");
      await expect(args.onChange).toHaveBeenNthCalledWith(2, "te");
      await expect(args.onChange).toHaveBeenNthCalledWith(3, "tes");
      await expect(args.onChange).toHaveBeenNthCalledWith(4, "test");
      await userEvent.tab();
    });
  },
};

export const DefaultValues: Story = {
  name: "Default Values",
  args: {
    value: "",
    name: "input",
    color: undefined,
    masked: undefined,
    status: undefined,
    disabled: undefined,
    placeholder: undefined,
    maxLength: undefined,
    type: undefined,
    onChange: fn(),
    onFocus: undefined,
    onBlur: undefined,
  },
};

export default meta;
