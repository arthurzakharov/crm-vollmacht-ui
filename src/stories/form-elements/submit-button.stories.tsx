import { Meta, StoryObj } from "@storybook/react";
import { SubmitButton } from "../../form-elements";
import { fn, within, expect, userEvent } from "@storybook/test";
import MaxWidthDecorator from "../decorators/max-width-decorator";

type Story = StoryObj<typeof SubmitButton>;

const meta: Meta<typeof SubmitButton> = {
  title: "FORM ELEMENTS/Submit Button",
  component: SubmitButton,
  decorators: [MaxWidthDecorator],
  parameters: {
    MaxWidthDecorator: 320,
  },
  args: {
    loading: false,
    text: "Submit button",
    onClick: fn(),
    version: "main",
  },
};

export const VersionSidebar: Story = {
  name: "Version Sidebar",
  args: {
    version: "sidebar",
  },
};

export const VersionMain: Story = {
  name: "Version Main",
  args: {
    version: "main",
  },
};

export const Loading: Story = {
  name: "Loading",
  args: {
    loading: true,
  },
  play: async ({ args, canvasElement, step }) => {
    const button = within(canvasElement).getByRole("button");
    await step("Button is disabled while loading", async () => {
      await userEvent.click(button);
      await expect(button).toBeDisabled();
      await expect(args.onClick).not.toBeCalled();
    });
  },
};

export const Idle: Story = {
  name: "Idle",
  args: {
    loading: false,
  },
  play: async ({ args, canvasElement, step }) => {
    const button = within(canvasElement).getByRole("button");
    await step("Button is not disable while idle", async () => {
      await userEvent.click(button);
      await expect(button).not.toBeDisabled();
      await expect(args.onClick).toHaveBeenNthCalledWith(1);
    });
    await step("Button is not in focus after click", async () => {
      await userEvent.click(button);
      await expect(args.onClick).toHaveBeenNthCalledWith(2);
      await expect(button).not.toHaveFocus();
    });
  },
};

export default meta;
