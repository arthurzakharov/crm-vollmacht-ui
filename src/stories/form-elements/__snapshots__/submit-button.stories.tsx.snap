// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FORM ELEMENTS/Submit Button Idle play-test 1`] = `
<div style="max-width: 320px;">
  <button type="button"
          tabindex="0"
          class="submit-button submit-button--idle"
  >
    Submit button
  </button>
</div>
`;

exports[`FORM ELEMENTS/Submit Button Loading play-test 1`] = `
<div style="max-width: 320px;">
  <button type="button"
          tabindex="0"
          disabled
          class="submit-button submit-button--loading"
  >
    <div class="submit-button__loader">
      <svg viewbox="0 0 24 30"
           class="loader loader--secondary"
      >
        <rect x="0"
              y="10"
              class="loader__rect"
        >
          <animate attributename="opacity"
                   attributetype="XML"
                   values="0.2; 1; .2"
                   begin="0s"
                   dur="0.6s"
                   repeatcount="indefinite"
          >
          </animate>
          <animate attributename="height"
                   attributetype="XML"
                   values="10; 20; 10"
                   begin="0s"
                   dur="0.6s"
                   repeatcount="indefinite"
          >
          </animate>
          <animate attributename="y"
                   attributetype="XML"
                   values="10; 5; 10"
                   begin="0s"
                   dur="0.6s"
                   repeatcount="indefinite"
          >
          </animate>
        </rect>
        <rect x="8"
              y="10"
              class="loader__rect"
        >
          <animate attributename="opacity"
                   attributetype="XML"
                   values="0.2; 1; .2"
                   begin="0.15s"
                   dur="0.6s"
                   repeatcount="indefinite"
          >
          </animate>
          <animate attributename="height"
                   attributetype="XML"
                   values="10; 20; 10"
                   begin="0.15s"
                   dur="0.6s"
                   repeatcount="indefinite"
          >
          </animate>
          <animate attributename="y"
                   attributetype="XML"
                   values="10; 5; 10"
                   begin="0.15s"
                   dur="0.6s"
                   repeatcount="indefinite"
          >
          </animate>
        </rect>
        <rect x="16"
              y="10"
              class="loader__rect"
        >
          <animate attributename="opacity"
                   attributetype="XML"
                   values="0.2; 1; .2"
                   begin="0.3s"
                   dur="0.6s"
                   repeatcount="indefinite"
          >
          </animate>
          <animate attributename="height"
                   attributetype="XML"
                   values="10; 20; 10"
                   begin="0.3s"
                   dur="0.6s"
                   repeatcount="indefinite"
          >
          </animate>
          <animate attributename="y"
                   attributetype="XML"
                   values="10; 5; 10"
                   begin="0.3s"
                   dur="0.6s"
                   repeatcount="indefinite"
          >
          </animate>
        </rect>
      </svg>
    </div>
    Submit button
  </button>
</div>
`;

exports[`FORM ELEMENTS/Submit Button VersionMain smoke-test 1`] = `
<div style="max-width: 320px;">
  <button type="button"
          tabindex="0"
          class="submit-button submit-button--idle"
  >
    Submit button
  </button>
</div>
`;

exports[`FORM ELEMENTS/Submit Button VersionSidebar smoke-test 1`] = `
<div style="max-width: 320px;">
  <button type="button"
          tabindex="0"
          class="submit-button submit-button--idle submit-button--sidebar-version"
  >
    Submit button
  </button>
</div>
`;
